transport/src/main/java/io/netty/channel/FileChannelBasedSlicedFileRegion.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * A derived {@link FileChannelBasedFileRegion} which shares reference counting with its parent.
+ */
+final class FileChannelBasedSlicedFileRegion extends FileChannelBasedFileRegion {
+
+    private final FileChannelBasedFileRegion parent;
+
+    private final long offset;
+
+    private final long count;
+
+    public FileChannelBasedSlicedFileRegion(FileChannelBasedFileRegion parent, long index,
+            long count) {
+        if (index < 0 || index > parent.count() - count) {
+            throw new IndexOutOfBoundsException(parent + ".slice(" + index + ", " + count + ')');
+        }
+
+        if (parent instanceof FileChannelBasedSlicedFileRegion) {
+            FileChannelBasedSlicedFileRegion sliced = (FileChannelBasedSlicedFileRegion) parent;
+            this.parent = sliced.parent;
+            this.offset = sliced.offset + index;
+        } else {
+            this.parent = parent;
+            this.offset = index;
+        }
+        this.count = count;
+    }
+
+    @Override
+    public long position() {
+        return parent.position() + offset;
+    }
+
+    @Override
+    public long transfered() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long transferred() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long count() {
+        return count;
+    }
+
+    @Override
+    public long transferTo(WritableByteChannel target, long position) throws IOException {
+        throw new UnsupportedOperationException();
```

Comment:
> These methods are expected to be removed in the future

So this argument feels like "because we're lazy," because the methods _are_ there today. "Let's not spend the effort of implementing them if they will be removed later." Which is a fair argument, but not by itself.

My implementation sketch would seem to indicate it isn't very hard to add support for the old method. Some edge cases might need to be checked, but that still seems to be pretty easy. So assuming my proposed implementation works, the "lazy" here seems to be quite extreme, as we don't want to add ~10 straight-forward lines to maintain compatibility. But I'd accept that maybe there's an issue with my implementation, which could dramatically change the cost/benefit.

> new implementation classes are not required to implement them.

This is a circular argument. I'm claiming that implementations should implement them, and that I don't see a strong reason why they shouldn't implement them.

> If you use the new methods introduced in SliceableFileRegion then you need to use the new transferBytesTo methods.

As I mention in the other comment, the problem is that these are two different "you"s. The code creating the instance is different than the one consuming it, otherwise they wouldn't need to communicate via an interface.